sematic/plugins/abstract_metrics_storage.py

```diff
@@ -0,0 +1,202 @@
+"""
+Module declaring the interface for metrics storage plug-ins.
+"""
+# Standard Library
+import abc
+import enum
+from dataclasses import dataclass, field
+from datetime import datetime
+from typing import List, Literal, Optional, Sequence, Tuple, Type, Union, cast
+
+# Sematic
+from sematic.abstract_plugin import AbstractPlugin, PluginScope
+from sematic.config.settings import get_active_plugins
+from sematic.metrics.metric_point import MetricPoint, MetricsLabels
+
+
+@dataclass
+class MetricsFilter:
+    """
+    Data structure to query metrics.
+    """
+
+    name: str
+    from_time: datetime
+    to_time: datetime
+    labels: MetricsLabels
+
+
+@dataclass
+class MetricSeries:
+    """
+    Data structure to represent a metric series for queries to return.
+
+    Series are typically aggregated by labels.
+
+    Parameters
+    ----------
+    metric_name: str
+        The name of the metric
+    metric_type: str
+        One of `MetricType`. Using the str value instead of Enum instance for
+        serializability.
+    series: List[Tuple[float, Tuple[str, ...]]]
+        The metric series. A list of tuples. Every element in the list
+        correspond to a series value. The tuple contains two elements: the
+        value, and a tuple of label values. The order of label values
+        corresponds to the order of label names in columns.
+    columns: List[str]
+        The list of label names the metric was aggregated over.
+
+    Examples
+    --------
+    Run count with no aggregations.
+    ```
+    MetricSeries(
+        metric_name="sematic.func_run_count",
+        metric_type="COUNT",
+        series=[(483.0, ())],
+        columns=[]
+    )
+    ```
+
+    Success rate by calculator path by day
+    ```
+    MetricSeries(
+        metric_name="sematic.func_success_rate",
+        metric_type="GAUGE",
+        series=[
+            (0.68, ("path.to.foo", "2023-04-12")),
+            (0.95, ("path.to.foo", "2023-04-11")),
+            (0.45, ("path.to.bar", "2023-04-12")),
+            ...
+        ],
+        columns=["calculator_path", "date"],
+    )
+    ```
+    """
+
+    metric_name: str
+    metric_type: Optional[str] = None
+    series: List[Tuple[float, Tuple[Union[str, int], ...]]] = field(
+        default_factory=list
+    )
+    columns: List[str] = field(default_factory=list)
```

Comment:
I would avoid "columns", as it would suggest this is backed by or referring to a DB table.
```suggestion
    aggregations: List[str] = field(default_factory=list)
```

Reply:
`columns` was suggested by @augray. It refers to the columns of the time series tuples. I think it's fine. Pandas dataframes have columns and do not refer to a DB table.